nidx/src/maintenance/scheduler.rs

```diff
@@ -158,23 +169,112 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
     Ok(())
 }
 
-pub async fn schedule_merges(meta: &NidxMetadata, consumer: &mut PullConsumer) -> anyhow::Result<()> {
-    let oldest_confirmed_seq = consumer.info().await?.ack_floor.stream_sequence;
-    let oldest_pending_seq = oldest_confirmed_seq + 1;
-
-    // Enqueue merges. Right now, merge everything that we can
+/// Enqueue merge jobs for segments older than `before_seq` that aren't already
+/// scheduled or marked to delete. Right now we merge everything that we can
+async fn schedule_merges(meta: &NidxMetadata, before_seq: Seq) -> anyhow::Result<()> {
     // TODO: better merge algorithm
-    let merges = sqlx::query_scalar!(
-        r#"SELECT array_agg(id) AS "segment_ids!" FROM segments WHERE delete_at IS NULL AND merge_job_id IS NULL AND seq < $1 GROUP BY index_id"#,
-        oldest_pending_seq as i64
+    let merges = sqlx::query!(
+        r#"SELECT index_id, array_agg(id) AS "segment_ids!: Vec<SegmentId>" FROM segments WHERE delete_at IS NULL AND merge_job_id IS NULL AND seq < $1 GROUP BY index_id"#,
+        i64::from(before_seq),
     )
     .fetch_all(&meta.pool)
     .await?;
     for m in merges {
-        if m.len() > 3 {
-            MergeJob::create(meta, &m, oldest_confirmed_seq as i64).await?;
+        if m.segment_ids.len() > 3 {
+            MergeJob::create(meta, m.index_id.into(), &m.segment_ids, before_seq).await?;
```

Comment:
You introduced a consistency bug here. Before, this was `oldest_confirmed_seq` and now it is `before_seq = oldest_pending_seq`.

An option maybe to change the call to this function to pass `oldest_confirmed_seq` and change the query to be `<=` instead of `<`.

**Explanation**
The key concept here is that we are merging segments while also indexing new ones. Merging also involves applying all deletions and since new segments may affect older segments (they include updates for the same resource in a previous segment), it is crucial that we merge "in-order". The situation we want to avoid is the following:
- We merge segments with seqs 100 and 102, applying all deletions. The seq number of the new segment is 102.
- We index segment seq 101 that includes deletions which affect 100. However, since 100 is merged into 102, it will never see the deletions for 100.

In order to ensure this, we only merge segments older than the oldest pending seq (nats message still processing). In this case, since message 101 is still pending (not indexed = not ack), NATS will report `oldest_confirmed_seq = consumer_info.ack_floor.stream_sequence = 100`. We can merge all segments up to and including 100 (< 101 in the previous code, it could be <= 100), and the resulting seq for the merged segment will be 100.

**Note**: Since this is super-important, it would be nice to have a explicit test for it.

Reply:
Fixed! Let me know what you think. I also added some docs and tests